packages/adminrouter/extra/src/README.md

```diff
@@ -31,6 +31,226 @@ Use `make check-api-docs` to validate that the YAML and HTML files are up to dat
+as it is the case for nginx includes and thus it's possible to simplify the code
+a bit. It is sufficient to just differentiate lua code basing on the repo flavour.
+Both agents and masters can share the same lua code.
+
+There are two possible reasons that may be preventing lua code from being
+shared:
+* the same code is executed but with different call arguments. An example to
+  this may be `auth.validate_jwt_or_exit()` function. In Open it takes no arguments,
+  in EE it takes more than one.
+* the code differs between ee and open but shares some common libraries/functions.
+  Great example for this is `auth.check_acl_or_exit()` which internally, among
+  many other things, calls argument-less `auth.check_jwt_or_exit()`.
+
+In order to address these issues, a couple of patterns were selected:
+ * modules which have flavour-specific function arguments export argument-less
+   functions for the NGINX configuration. They translate the original call into
+   a call with correct arguments. This approach requires splitting module into
+   `ee.lua/open.lua/common.lua` parts which is described in next bulletpoint.
+   Depending on the flavour, either `ee.lua` or `open.lua` is imported and correct
+   argument-less function is used. This approach also enables to share some of
+   the NGINX `location` blocks between flavours - `location` code is the same,
+   even though the Lua code used by the block differs.
+ * some libraries/functions are structured in a way that extracts parts common for
+   both flavours and `ee/open` parts that are included only in EE and Open repos
+   respectively. An example of this approach is auth library which is splitted
+   into three parts:
+	* `lib/auth/common.lua` - present only in Open repository, with functions
+	  shared by both EE and Open code.
+	* `lib/auth/ee.lua` - present only in EE repository, with functions specific
+      to EE that use boilerplate from lib/auth/common.lua.
+	* `lib/auth/open.lua` - as above but for Open repository.
+   `init_by_lua` OpenResty call in `includes/http/(open|ee)/common.conf` imports
+   auth.open or auth.ee modules as auth respectively.  Module is registered in
+   global namespace so all other lua code uses it. This approach also allows
+   for some degree of code separation enforcement as Open lua libs are removed
+   during EE repo apply.
+
+A special case of same code is executed but with different call arguments
+problem is when the module needs to be initialized differently/passed
+different arguments depending on the flavour. Great example of it is
+`cache.lua` module which in Open does not require any extra
+authentication data. In order to solve it, modules are required to provide
+`.init()` function which is returned by require `<module-name>` stetement.
+The init function accepts arguments which reconfigure module instance according
+to the flavour requirements.
+
+Modules that do not require any customization in regards of flavour are left
```

Comment:
Consider adding some examples: one example of a module that uses init() and another that does not.